sdk/metric/meter.go

```diff
@@ -19,21 +19,89 @@ package metric // import "go.opentelemetry.io/otel/sdk/metric"
 
 import (
 	"context"
+	"sync"
 
 	"go.opentelemetry.io/otel/metric"
 	"go.opentelemetry.io/otel/metric/instrument"
 	"go.opentelemetry.io/otel/metric/instrument/asyncfloat64"
 	"go.opentelemetry.io/otel/metric/instrument/asyncint64"
 	"go.opentelemetry.io/otel/metric/instrument/syncfloat64"
 	"go.opentelemetry.io/otel/metric/instrument/syncint64"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
 )
 
+// meterRegistry keeps a record of initialized meters for instrumentation
+// libraries. A meter is unique to an instrumentation library and if multiple
+// requests for that meter are made a meterRegistry ensure the same instance
+// is used.
+//
+// The zero meterRegistry is empty and ready for use.
+//
+// A meterRegistry must not be copied after first use.
+//
+// All methods of a meterRegistry are safe to call concurrently.
+type meterRegistry struct {
+	sync.Mutex
+
+	ordered []*meter
+	indexed map[instrumentation.Library]*meter
+}
+
+// Get returns a registered meter matching the instrumentation library if it
+// exists in the meterRegistry. Otherwise, a new meter configured for the
+// instrumentation library is registered and then returned.
+//
+// The returned found bool is true if the returned meter already existed in
+// the registry and false if a new meter was created.
+//
+// Get is safe to call concurrently.
+func (r *meterRegistry) Get(l instrumentation.Library) (m *meter, found bool) {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.indexed == nil {
+		r.indexed = make(map[instrumentation.Library]*meter, 1)
+	}
+
+	m, ok := r.indexed[l]
+	if ok {
+		return m, true
+	}
+
+	m = &meter{Library: l}
+	r.indexed[l] = m
+	r.ordered = append(r.ordered, m)
+
+	return m, false
+}
+
+// Range calls f sequentially for each meter present in the meterRegistry (in
+// the order they were added to the registry). If f returns false, the
```

Comment:
is the ordering important?

Reply:
Hmm, good question. I took the concept from https://github.com/open-telemetry/opentelemetry-go/pull/2865, but I don't see a computational reason it needs to be. I'll remove it.